kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/PartitionedDatasetWriter.java

```diff
@@ -259,4 +237,49 @@ public void onRemoval(
     }
 
   }
+
+  private static class IncrementalPartitionedDatasetWriter<E> extends
+      PartitionedDatasetWriter<E> implements org.kitesdk.data.Flushable, Syncable {
+
+    private IncrementalPartitionedDatasetWriter(FileSystemView<E> view) {
+      super(view);
+    }
+
+    @Override
+    public void flush() {
+      Preconditions.checkState(state.equals(ReaderWriterState.OPEN),
+          "Attempt to flush a writer in state:%s", state);
+
+      LOG.debug("Flushing all cached writers for view:{}", view);
+
+    /*
+     * There's a potential for flushing entries that are created by other
+     * threads while looping through the writers. While normally just wasteful,
+     * on HDFS, this is particularly bad. We should probably do something about
+     * this, but it will be difficult as Cache (ideally) uses multiple
+     * partitions to prevent cached writer contention.
+     */
+      for (FileSystemWriter<E> writer : cachedWriters.asMap().values()) {
+        LOG.debug("Flushing partition writer:{}", writer);
+        if (writer instanceof Flushable) {
```

Comment:
Good point. Yes, I think it is an error. It would be better if the cache was parameterized with IncrementalWriter for the incremental case to avoid type checking, even if it is inserted automatically with a cast.